### PR TITLE
Added api (mostly)

### PR DESCRIPTION
**Pull request by sirherrbatka**

### CODE REVIEW

File: src/utils/lists.lisp [hunk 1]
```diff
@@ -37,21 +37,16 @@
...
+    (for elt on list)
+    (when (funcall test
+                   (funcall key (car elt))
+                   item)
+      (return-from try-find-cell elt))))
 
 
 (defun try-find (item list &key (test #'eql) (key #'identity))
-  (multiple-value-bind (r f) (try-find-cell item list :test test :key key)
+  (multiple-value-bind (r) (try-find-cell item list :test test :key key)
```
Comment (MatthewRock) [inline src/utils/lists.lisp]: If only one value is returned, there is no need for multiple-value-bind

File: cl-data-structures.asd [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+(asdf:defsystem cl-data-structures
+  :name "cl-data-structures"
+  :version "0.0.0"
+  :license "MIT"
+  :author "Lisp Mechanics"
+  :maintainer "Lisp Mechanics"
+  :depends-on (:iterate :alexandria :serapeum)
+  :serial T
+  :components ((:file "package")
```
Comment (MatthewRock) [inline cl-data-structures.asd:11]: src should be added as system's pathname. Module name should be a string; it serves then as a pathname to the subsystem(so, cl-data-structures system's pathname "src" and module name "api" would imply that the pathname of its components is "src/api").

File: package.lisp [hunk 3]
```diff
@@ -0,0 +1,22 @@
+(defpackage :cl-data-structures
+  (:use :common-lisp)
```
Comment (MatthewRock) [inline package.lisp:2]: package should be inside src folder; exported symbols do not need to be quoted nor keyworded.

File: src/api/generics.lisp [hunk 4]
```diff
@@ -0,0 +1,65 @@
+(in-package :cl-data-structures)
+
+
+(defgeneric at (container location)
```
Comment (MatthewRock) [inline src/api/generics.lisp:4]: IMHO at should always return either one or two values to be consistent. Third option is to provide some keyword argument like :default which will be returned when nothing is found. We can discuss preference later.

File: src/api/generics.lisp [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  (:documentation "Obtain element stored at LOCATION in the CONTAINER. This function will return one or two values, depending on the CONTAINER.
+  In case of associative containers, second value informs if element was found (first value is nil if element was not found).
+  In case of non-associtive containers (e.g. vectors), the function returns value under LOCATION if LOCATION is valid, otherwise condition of type TBD will be raised."))
+
+
+(defgeneric (setf at) (new-value container location)
+  (:documentation "Destructively insert/replace element in the CONTAINER at LOCATION witn NEW-VALUE."))
```
Comment (MatthewRock) [inline src/api/generics.lisp:11]: s/witn/with

File: src/api/generics.lisp [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+
+(defgeneric add (container location new-value)
+  (:documentation "Non-destructively insert NEW-VALUE into CONTAINER at LOCATION. Will return two values: first one is the new container, second one is boolean informing if insert took place."))
+
+
+(defgeneric add! (container location new-value)
```
Comment (MatthewRock) [inline src/api/generics.lisp:18]: When would insert not take place?

File: src/api/generics.lisp [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  (:documentation "Destructively insert NEW-VALUE into CONTAINER at LOCATION. Will return two values: first one is the modified CONTAINER, second one is boolean informing if insert took place "))
+
+
+(defgeneric insert (container location new-value)
+  (:documentation "Non-destructively insert NEW-VALUE into CONTAINER at LOCATION. Will replace element value at LOCATION if it was already occupied. Will return up to three values: new container, boolean to inform user if element was already in the container and old value AT location (or nil if it was not present). Essentially purely functional (SETF (AT CONTAINER) NEW-VALUE)."))
+
+
+(defgeneric erase (container location)
```
Comment (MatthewRock) [inline src/api/generics.lisp:26]: Returns?

File: src/api/generics.lisp [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+(defgeneric add (container location new-value)
+  (:documentation "Non-destructively insert NEW-VALUE into CONTAINER at LOCATION. Will return two values: first one is the new container, second one is boolean informing if insert took place."))
+
+
+(defgeneric add! (container location new-value)
+  (:documentation "Destructively insert NEW-VALUE into CONTAINER at LOCATION. Will return two values: first one is the modified CONTAINER, second one is boolean informing if insert took place "))
+
+
+(defgeneric insert (container location new-value)
```
Comment (MatthewRock) [inline src/api/generics.lisp:22]: I think the amount of returned values should be constant.

File: src/api/generics.lisp [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+(defgeneric update (container location new-value)
+  (:documentation "If LOCATION is taken in the CONTAINER, update it. Returns three values: first: new CONTAINER with update LOCATION, second: t if update took place; nil otherwise, previous value."))
+
+
+(defgeneric update! (new-value container location)
+  (:documentation "Destructive version of UPDATE."))
+
+
+(defgeneric become-functional (container)
```
Comment (MatthewRock) [inline src/api/generics.lisp:46]: What is the purpose of this generic function?

File: src/api/generics.lisp [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  (:method ((container functional)) container))
+
+
+(defgeneric become-mutable (container)
```
Comment (MatthewRock) [inline src/api/generics.lisp:50]: What is the purpose of this generic function?

File: src/api/generics.lisp [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+  (:method ((container mutable)) t)
+  (:method ((container fundamental-container)) nil))
+
+
+(defgeneric functional-p (container)
+  (:method ((container functional)) t)
+  (:method ((container fundamental-container)) nil))
+
+
+(defgeneric empty (container)
```
Comment (MatthewRock) [inline src/api/generics.lisp:64]: I think the name should be empty-p or something

File: src/utils/modification-algorithms.lisp [hunk 12]
```diff
@@ -0,0 +1,32 @@
+(in-package :cl-ds.utils)
+
+
+(-> lastpop ((and vector (not simple-array))) t)
+(defun lastpop (vector)
```
Comment (MatthewRock) [inline src/utils/modification-algorithms.lisp:5]: Maybe pop-last ?

File: src/utils/modification-algorithms.lisp [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    (for index = (- el i))
+    (replace vector vector
+             :start1 index
+             :start2 (1+ index))
+    (finally (progn (decf (fill-pointer vector) (1+ i))
+                    (return vector)))))
+
+
+(-> swapop ((AND VECTOR (NOT SIMPLE-ARRAY)) index) vector)
+(defun swapop (vector index)
```
Comment (MatthewRock) [inline src/utils/modification-algorithms.lisp:25]: Maybe swap-and-pop-last ? :)

File: src/utils/modification-algorithms.lisp [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+(-> lastpop ((and vector (not simple-array))) t)
+(defun lastpop (vector)
+  (unless (zerop (fill-pointer vector))
+    (prog1
+        (aref vector (1- (fill-pointer vector)))
+      (decf (fill-pointer vector)))))
+
+
+(-> erase-from-vector (vector list) vector)
+(defun erase-from-vector (vector indexes)
```
Comment (MatthewRock) [inline src/utils/modification-algorithms.lisp:13]: This function could use documentation string

File: src/utils/macros.lisp [hunk 15]
```diff
@@ -0,0 +1,138 @@
+(in-package :cl-data-structures.utils)
+
```
Comment (MatthewRock) [inline src/utils/macros.lisp:2]: Macros/functions below could use documentation

File: src/utils/lists.lisp [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+        ((t nil) (progn (push element result)
+                        (setf last-cell result
+                              (cdr last-cell) (cdr rest))
+                        (return-from insert-or-replace (values result t elt))))
+        ((t t) (progn (push element result)
+                      (setf (cdr last-cell) (cdr rest))
+                      (return-from insert-or-replace (values result t elt))))))
+    (values (cons element set) nil nil)))
+
+
```
Comment (MatthewRock) [inline src/utils/lists.lisp:25]: functions below could use documentation

File: src/utils/package.lisp [hunk 17]
```diff
@@ -0,0 +1,22 @@
+(defpackage :cl-data-structures.utils
+  (:use :common-lisp :iterate :alexandria :serapeum)
+  (:nicknames :cl-ds.utils)
+  (:shadowing-import-from :iterate :collecting :summing :in)
+  (:export
```
Comment (MatthewRock) [inline src/utils/package.lisp:5]: exported symbols do not need to be quoted or keyworded

**Review by MatthewRock — CHANGES_REQUESTED** [review]
I made a review. It would be nice if you could look at the comments.

**Comment by MatthewRock** [comment]
Could you please cover some of the code with tests?

File: src/api/generics.lisp (hunk 4 above)
Comment (sirherrbatka) [reply src/api/generics.lisp:4]: Well, vectors can just return (values result t) i guess.

File: package.lisp (hunk 3 above)
Comment (sirherrbatka) [reply package.lisp:2]: Why?

File: cl-data-structures.asd (hunk 2 above)
Comment (sirherrbatka) [reply cl-data-structures.asd:11]: I don't see the point. cl-ds.api vs simply cl-ds seems to be just redundant.

File: src/utils/lists.lisp (hunk 16 above)
Comment (sirherrbatka) [reply src/utils/lists.lisp:25]: sure they could…

File: src/utils/lists.lisp (hunk 1 above)
Comment (sirherrbatka) [reply src/utils/lists.lisp]: that is actually good point.

File: src/api/generics.lisp (hunk 9 above)
Comment (sirherrbatka) [reply src/api/generics.lisp:46]: Change interface from mutable to functional.

File: src/api/generics.lisp (hunk 10 above)
Comment (sirherrbatka) [reply src/api/generics.lisp:50]: Change interface from functional to mutable.

File: src/api/generics.lisp (hunk 11 above)
Comment (sirherrbatka) [reply src/api/generics.lisp:64]: Dunno. CL is not that consintent here.

File: src/utils/modification-algorithms.lisp (hunk 12 above)
Comment (sirherrbatka) [reply src/utils/modification-algorithms.lisp:5]: Maybe?

File: src/utils/modification-algorithms.lisp (hunk 13 above)
Comment (sirherrbatka) [reply src/utils/modification-algorithms.lisp:25]: That's way to long.

**Comment by sirherrbatka** [comment]
Eventually I will add tests (i have some, but i want to write generic tests that can be reusable for all dicts). But i ran out of time for now.

**Pull request closed without merging**